source/common/protobuf/deterministic_hash.cc

```diff
@@ -10,12 +10,12 @@ namespace {
 
 #define REFLECTION_FOR_EACH(get_type, value_type)                                                  \
   if (field->is_repeated()) {                                                                      \
-    for (const auto q : reflection->GetRepeatedFieldRef<value_type>(message, field)) {             \
+    for (const value_type& q : reflection->GetRepeatedFieldRef<value_type>(message, field)) {      \
       seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)},   \
                                 seed);                                                             \
     }                                                                                              \
   } else {                                                                                         \
-    const auto q = reflection->Get##get_type(message, field);                                      \
+    const value_type q = reflection->Get##get_type(message, field);                                \
```

Comment:
is there a reference version of Get? This would deep-copy a string, right?

Reply:
Yes, that's why it's not used for string.

Reply:
what about other messages. what if this is a repeated complex message?

Reply:
Those don't go into this code either. Now made explicit by template so you *can't* get strings or messages into the scalar hashing function.